system/services/admin/DataManagerBatchOperationService.cfc

```diff
@@ -0,0 +1,429 @@
+		}
+
+		try {
+			do {
+				if ( arguments.batchAll ) {
+					sourceIds = getNextBatchRecordsFromQueue( queueId, 100 );
+
+					if ( !ArrayLen( sourceIds ) ) {
+						break;
+					}
+
+					if ( canInfo ) {
+						arguments.logger.info( $translateResource( uri="cms:datamanager.batchedit.fetched.records", data=[ NumberFormat( ArrayLen( sourceIds ) ) ] ) );
+					}
+
+					moreToFetch = ArrayLen( sourceIds ) == 100;
+				}
+
+				for( var sourceId in sourceIds ) {
+					if ( $isInterrupted() ) {
+						if ( canWarn ) { arguments.logger.warn( "Task interrupted. Cancelling." ); }
+						break;
+					}
+					if ( !isMultiValue ) {
+						pobjService.updateData(
+							  objectName = objectName
+							, data       = { "#arguments.fieldName#" = value }
+							, filter     = { id=sourceId }
+						);
+					} else {
+						var existingIds  = [];
+						var targetIdList = [];
+						var newChoices   = ListToArray( arguments.value );
+
+						if ( arguments.multiEditBehaviour != "overwrite" ) {
+							var previousData = pobjService.getDeNormalizedManyToManyData(
+								  objectName   = objectName
+								, id           = sourceId
+								, selectFields = [ arguments.fieldName ]
+							);
+							existingIds = ListToArray( previousData[ arguments.fieldName ] ?: "" );
+						}
+
+						switch( arguments.multiEditBehaviour ) {
+							case "overwrite":
+								targetIdList = newChoices;
+								break;
+							case "delete":
+								targetIdList = existingIds;
+								for( var id in newChoices ) {
+									targetIdList.delete( id )
+								}
+								break;
+							default:
+								targetIdList = existingIds;
+								targetIdList.append( newChoices, true );
+						}
+
+						targetIdList = targetIdList.toList();
+						targetIdList = ListRemoveDuplicates( targetIdList );
+
+						pobjService.updateData(
+							  objectName              = objectName
+							, id                      = sourceId
+							, data                    = { "#arguments.fieldName#" = targetIdList }
+							, updateManyToManyRecords = true
+						);
+					}
+
+					$audit(
+						  action   = arguments.auditAction
+						, type     = arguments.auditCategory
+						, userId   = arguments.auditUserId
+						, recordId = sourceid
+						, detail   = Duplicate( arguments )
+					);
+
+					if ( canReportProgress ) {
+						arguments.progress.setProgress( Int( ( 100 / totalrecords ) * ++processed ) ) ;
+					}
+				}
+
+				if ( $isInterrupted() ) {
+					break;
+				}
+			} while( moreToFetch );
+		} catch( any e ) {
+			if ( Len( queueId ) ) {
+				clearBatchOperationQueue( queueId );
+			}
+
+			rethrow;
+		}
+
+		if ( Len( queueId ) ) {
+			clearBatchOperationQueue( queueId );
+		}
+		if ( canInfo ) {
+			arguments.logger.info( $translateResource( uri="cms:datamanager.batchedit.task.finished.message", data=[ objectTitle, fieldTitle, NumberFormat( totalRecords ) ] ) );
```

Comment:
Is totalRecords right in the finished.message? The task could have been interrupted before that number of records has been processed